File: policies/isSingapore/swift/Tests/IsSingaporeTests.swift
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Understone, Inc.

import XCTest
import OctetSDK
@testable import IsSingapore
@testable import OctetPolicyCore

/// Tests for `isSingapore` against the real region-containment model.
/// The predicate's testable core takes an `evaluate` closure, so each
/// test injects a canned `OctetVerdict` for the region the predicate
/// queries — no live SDK or license key needed.
final class IsSingaporeTests: XCTestCase {

    private let fixedTime = Date(timeIntervalSince1970: 1_716_000_000)

    // 0.0.1-alpha LocationProof has no default init; tests don't read
    // proof bytes, so synthesize a placeholder with arbitrary fields.
    private func fakeProof() -> LocationProof {
        LocationProof(
            id: "test", claimedRegion: .country(isoCode: "SG"),
            level: .country, timestampMs: 0,
            confidence: ConfidenceSummary(),
            positionCommitment: Data(), sdkVersion: "test",
            platform: "test", proofBytes: Data()
        )
    }

    /// Build a verdict honoring the SDK's invariant: `proof` is nil
    /// iff `result == .indeterminate`.
    private func verdict(
        _ result: OctetVerdict.Result,
        reason: OctetVerdict.ReasonCode = .ok
    ) -> OctetVerdict {
        let proof: LocationProof? = (result == .indeterminate) ? nil : fakeProof()
        return OctetVerdict(
            result: result,
            reason: reason,
            message: "test",
            proof: proof,
            validity: nil,
            queriedAt: fixedTime,
            confidence: ConfidenceSummary()
        )
    }

    func test_device_in_singapore_matches() async {
        // The SDK confirms containment in SG → match.
        let result = await isSingapore { region in
            XCTAssertEqual(region.shape, .country(isoCode: "SG"),
                           "isSingapore must query exactly the SG country region")
            return self.verdict(.yes)
        }

        XCTAssertTrue(result.match)
        XCTAssertEqual(result.country, "SG")
        XCTAssertNil(result.state)
        XCTAssertEqual(result.reason, .ok)
        XCTAssertEqual(result.confidence, .high)
        XCTAssertEqual(result.policyName, "isSingapore")
        XCTAssertEqual(result.policyVersion, PolicyPackage.version)
        XCTAssertEqual(result.evaluatedAt, fixedTime)
    }

    func test_device_not_in_singapore_is_clean_negative() async {
        let result = await isSingapore { _ in self.verdict(.no) }

        XCTAssertFalse(result.match)
        XCTAssertNil(result.country,
                     "a NO verdict doesn't reveal where the device is — country stays nil")
        XCTAssertEqual(result.reason, .ok)
    }

    func test_indeterminate_verdict_is_unknown_not_negative() async {
        let result = await isSingapore { _ in self.verdict(.indeterminate, reason: .noFix) }

        XCTAssertFalse(result.match)
        XCTAssertNil(result.country)
        XCTAssertEqual(result.reason, .verdictIndeterminate,
                       "match=false here means 'cannot tell', not 'device is not in SG'")
    }

    func test_not_yet_released_geometry_is_indeterminate() async {
        // If the SDK build can't evaluate the region yet it returns
        // INDETERMINATE / notYetReleased. The predicate folds that into
        // .verdictIndeterminate — still "unknown", never a false negative.
        let result = await isSingapore { _ in
            self.verdict(.indeterminate, reason: .notYetReleased)
        }

        XCTAssertFalse(result.match)
        XCTAssertEqual(result.reason, .verdictIndeterminate)
    }

    func test_evaluated_at_copies_through_from_verdict() async {
        let result = await isSingapore { _ in self.verdict(.yes) }
        XCTAssertEqual(result.evaluatedAt, fixedTime)
    }
}

File: policies/isSingapore/kotlin/src/test/kotlin/com/octetproof/octetpolicy/IsSingaporeTests.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Understone, Inc.

package com.octetproof.octetpolicy

import com.octetproof.sdk.api.ConfidenceSummary
import com.octetproof.sdk.api.LocationProof
import com.octetproof.sdk.api.OctetRegion
import com.octetproof.sdk.api.OctetVerdict
import com.octetproof.sdk.api.ProofLevel
import com.octetproof.sdk.api.ProofRegion
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.time.Instant

/**
 * Tests for `isSingapore` against the real region-containment model.
 * The testable core takes a `suspend (OctetRegion) -> OctetVerdict`
 * closure, so each test injects a canned verdict — no live SDK or
 * license key needed. Mirrors the Swift IsSingaporeTests.
 */
class IsSingaporeTests {

    private val fixedTime: Instant = Instant.ofEpochSecond(1_716_000_000)

    // 0.0.1-alpha LocationProof has no default instance; tests don't read
    // proof bytes, so synthesize a placeholder with arbitrary fields.
    private fun fakeProof(): LocationProof = LocationProof(
        "test", ProofRegion.Country("SG"), ProofLevel.COUNTRY, 0L,
        ConfidenceSummary(), ByteArray(0), "test", "test", ByteArray(0),
    )

    /** Build a verdict honoring the SDK invariant: proof is null iff
     *  result == INDETERMINATE. */
    private fun verdict(
        result: OctetVerdict.Result,
        reason: OctetVerdict.ReasonCode = OctetVerdict.ReasonCode.OK,
    ): OctetVerdict {
        val proof: LocationProof? =
            if (result == OctetVerdict.Result.INDETERMINATE) null
            else fakeProof()
        return OctetVerdict(
            result = result,
            reason = reason,
            message = "test",
            proof = proof,
            validity = null,
            queriedAt = fixedTime,
            confidence = ConfidenceSummary(),
        )
    }

    @Test
    fun `device in singapore matches`() = runTest {
        val result = isSingapore { region ->
            assertEquals(OctetRegion.country("SG"), region,
                "isSingapore must query the SG country region")
            verdict(OctetVerdict.Result.YES)
        }
        assertTrue(result.match)
        assertEquals("SG", result.country)
        assertNull(result.state)
        assertEquals(PolicyReasonCode.OK, result.reason)
        assertEquals(Confidence.HIGH, result.confidence)
        assertEquals("isSingapore", result.policyName)
        assertEquals(PolicyPackage.VERSION, result.policyVersion)
        assertEquals(fixedTime, result.evaluatedAt)
    }

    @Test
    fun `device not in singapore is clean negative`() = runTest {
        val result = isSingapore { verdict(OctetVerdict.Result.NO) }
        assertFalse(result.match)
        assertNull(result.country,
            "a NO verdict doesn't reveal where the device is — country stays null")
        assertEquals(PolicyReasonCode.OK, result.reason)
    }

    @Test
    fun `indeterminate verdict is unknown not negative`() = runTest {
        val result = isSingapore {
            verdict(OctetVerdict.Result.INDETERMINATE, OctetVerdict.ReasonCode.NO_FIX)
        }
        assertFalse(result.match)
        assertNull(result.country)
        assertEquals(PolicyReasonCode.VERDICT_INDETERMINATE, result.reason)
    }

    @Test
    fun `not yet released geometry folds to indeterminate`() = runTest {
        val result = isSingapore {
            verdict(OctetVerdict.Result.INDETERMINATE, OctetVerdict.ReasonCode.NOT_YET_RELEASED)
        }
        assertFalse(result.match)
        assertEquals(PolicyReasonCode.VERDICT_INDETERMINATE, result.reason)
    }
}
